[website] Add CPU quantization tutorial (#20856)
apache/incubator-mxnet: docs/python_docs/python/tutorials/performance/backend/dnnl/dnnl_quantization.md

    def forward(self, x):
        out = mx.npx.activation(self.bn1(self.conv1(x)), 'relu')
        out = self.bn2(self.conv2(out))
        out = mx.npx.activation(out + x, 'relu')
        return out
        
net = SampleBlock()
net.initialize()

data = mx.np.zeros(shape=(1,64,224,224))
# run fusion
net.optimize_for(data, backend='ONEDNN')

# We can check fusion by plotting current symbol of our optimized network
sym, _ = net.export(None)
graph = mx.viz.plot_network(sym, save_format='jpg')
graph.view()
```
Both HybridBlock and Symbol classes provide API to easily run fusion of operators. Single line of code is enabling fusion passes on model:
```
net.optimize_for(data, backend='ONEDNN')
```

*optimize_for* function is available also as Symbol class method. Example call to this API is shown below. Notice that Symbol’s *optimize_for* method is not done in-place, so assigning it to a new variable is required:

```
optimized_symbol = sym.optimize_for(backend='ONEDNN')
```

For the above model definition in a naive benchmark with artificial data, we can gain up to 1.75x speedup without any accuracy loss on our testing machine with Intel(R) Core(TM) i9-9940X.

> ## Quantization

Reviewer: Inconsistent new line after ## and ### - seems to be not too important as in HackMD preview it does look identical with or without it.

Author: Done

                                  transforms.ToTensor(),
                                  transforms.Normalize(mean=rgb_mean, std=rgb_std)])
val_data = mx.gluon.data.DataLoader(dataset.transform_first(transformer), batch_size, shuffle=True)

net = resnet50_v1(pretrained=True)
net.hybridize(backend='ONEDNN', static_alloc=True, static_shape=True)

top1, top5 = test_accuracy(net, val_data)
print('FP32 Top1 Accuracy: {} Top5 Accuracy: {}'.format(top1, top5))

qnet = quantize_net(net, calib_mode='naive', calib_data=val_data, num_calib_batches=10)
qnet.hybridize(static_alloc=True, static_shape=True)
top1, top5 = test_accuracy(qnet, val_data)
print('INT8Naive Top1 Accuracy: {} Top5 Accuracy: {}'.format(top1, top5))

qnet = quantize_net(net, calib_mode='entropy', calib_data=val_data, num_calib_batches=10)
qnet.hybridize(static_alloc=True, static_shape=True)
top1, top5 = test_accuracy(qnet, val_data)
print('INT8Entropy Top1 Accuracy: {} Top5 Accuracy: {}'.format(top1, top5))
```

#### Output:
> FP32 Top1 Accuracy: 0.76364 Top5 Accuracy: 0.93094
> INT8Naive Top1 Accuracy: 0.76028 Top5 Accuracy: 0.92796
> INT8Entropy Top1 Accuracy: 0.76404 Top5 Accuracy: 0.93042

> With quantized model there is tiny accuracy drop, however this is cost of great performance optimization and memory footprint. The difference between calibration method is dependent on the model itself, used activation layers and the size of calibration data.

Reviewer: ```suggestion
With quantized model there is tiny accuracy drop, however this is the cost of great performance optimization and memory footprint reduction. The difference between calibration methods is dependent on the model itself, used activation layers and the size of calibration data.
```

### Custom layer collectors and calibrating the model
In MXNet 2.0 new interface for creating custom calibration collector has been added. Main goal of this interface is to give the user as much flexibility as possible in almost every step of quantization. Creating own layer collector is pretty easy, however computing effective min/max values can be not a trivial task. 

> Layer collectors are responsible for collecting statistics of each node in the graph — it means that the input/output data of every operator executed can be observed. Collector utilize the register_op_hook method of HybridBlock class.

Reviewer: ```suggestion
Layer collectors are responsible for collecting statistics of each node in the graph — it means that the input/output data of every operator executed can be observed. Collector utilizes the register_op_hook method of HybridBlock class.
```

### Example of usage with quantization API:
```
from mxnet.contrib.quantization import *
import logging
logging.basicConfig(level=logging.DEBUG)

#…

calib_data_loader = mx.gluon.data.DataLoader(…)
my_collector = ExampleNaiveCollector(logger=logging.getLogger())

qnet = quantize_net(net, calib_mode='custom', calib_data=calib_data_loader, LayerOutputCollector=my_collector)
```

## Performance and accuracy results

### Performance
Performance results of CV models. Chart presents three different runs: base float32 model without optimizations, fused float32 model with optimizations and quantized model.
![performance](https://github.com/dmlc/web-data/blob/main/mxnet/tutorials/onednn/quantization_2_0/speedup.png?raw=true)
###### Relative Inference Performance (img/sec) for Batch Size 128

### Accuracy
Accuracy results of CV models. Chart presents three different runs: base float32 model without optimizations, fused float32 model with optimizations and quantized model.
![accuracy](https://github.com/dmlc/web-data/blob/main/mxnet/tutorials/onednn/quantization_2_0/accuracy.png?raw=true)
###### ImageNet(ILSVRC2012) TOP1 validation accuracy

> ##Notes

Reviewer: ```suggestion
## Notes
```
